### ISSUE TITLE
Support for scipy sparse matrices

### ISSUE THREAD

**Issue by dar326** [bayespy/bayespy#88]
Would it be possible to add support for scipy sparse matrices when observing variables. I have a situation where the state space for an observed (multinomial) node is large but the number of states observed for any observation will be much smaller than the size of the state space. Currently, observing data requires specifying a value for all possible states but it would be useful to not have to specify states with a 0 value. 

Perhaps, there is something about the structure of the code which makes this difficult or impossible to support but if its possible I could significantly reduce memory usage in my project.

Thanks!

**Comment by jluttine** [comment]
Yep, that's a good idea. However, at least some years ago sparse matrices were buggy and caused a lot of pain because some NumPy functions didn't work out of the box but one needed to write if-clauses and special treatment for sparse matrices. I don't know what's the status now. Also, at that time, SciPy supported only sparse *matrices*, not *arrays*. I think that hasn't changed. That's a bit of a problem because BayesPy is based on supporting arrays (e.g., one can specify multiple plate axes). So, the support would be limited to matrices only, though that would suffice for your case, I suppose.

Another possible solution is to store the observed state index instead of zeros for all other states and one for the observed state. However, this leads to different handling of observed and latent multinomial nodes. Also, it would be a bit difficult to support multinomial nodes that are partially observed.

So neither solution is perfect but they have their advantages, and I suppose either one would solve your problem. I need to think about this a little bit.

**Comment by jluttine** [comment]
Note to self: Add support in such a way that if the observation array is given as a sparse scipy matrix, then it will stay that way without conversion to a dense array.

**Comment by dar326** [comment]
Thanks for considering the addition of this feature. I guess if a library doesn't strictly follow the numpy array API, such a change would require many more changes throughout the BayesPy code base. So, a library such as this one wouldn't be ideal:

https://pythonhosted.org/sppy/csarray.html

**Comment by jluttine** [comment]
For your information, I looked into this during the weekend. I didn't quite finish it yet, perhaps 2/3 done, but I think sparse matrix support can be added. But it's a bit fragile because of the extremely bad sparse matrix compatibility with NumPy. Anyway, I hope I find time to finish it soon.

**Comment by dar326** [comment]
That is awesome. Thanks! Looking forward to seeing this feature

**Comment by jluttine** [comment]
I implemented very limited support for sparse matrices in Multinomial and Mixture nodes. It's available in the same branch as before:  `feature/backpropagation-messaging`. This feature is useful if you have a large number of samples from a high-dimensional Multinomial mixture distribution with only a small number of non-zero observations.

First, disclaimer: As I said, sparse matrix support in NumPy/SciPy is very bad and now I believe so even more. For instance, some operations with sparse matrices return weird arrays with sparse matrices as object elements, if the operation doesn't support sparse matrices. Also, I reported a bug to SciPy that sparse and dense element-wise multiplication results in a dense array instead of sparse matrix when broadcasting occurs. Also, no support for sparse arrays/tensors, and numpy.einsum doesn't work with sparse matrices. And so on, and so on, the list goes on... So, if you try to play with sparse matrices and end up seeing something weird you don't understand, it's probably related to these issues.

For these reasons, the support is rather fragile, and the model must be created in a specific way. Below is a working example, I hope you can modify that to your use case. But the main point I'd like to emphasize is that `cluster_plate` must be set to `-2`, that is, the second last plate axis in the multinomial *parameters* is the cluster axis.


```python
import numpy as np
from scipy import sparse
import bayespy as bp

np.random.seed(42)

# Dimensionality of the multinomial vectors
D = 100

# Number samples
N = 1000

# Number of estimated clusters
K = 10

# Number of true clusters
K_true = 3

# Number of trials in the multinomials
L = 20

# Prior cluster assignment probabilities
alpha = bp.nodes.Dirichlet(np.ones(K))

# Unknown cluster assignments
Z = bp.nodes.Categorical(alpha, plates=(N,))

# Probability vector for each cluster.
# NOTE: because cluster plate axis is the second last axis, we need to add a singleton
# plate axis to the end
p = bp.nodes.Dirichlet(1e0*np.ones(D), plates=(K, 1))

# Multinomial observations. NOTE: cluster_plate must be -2!
Y = bp.nodes.Mixture(Z, bp.nodes.Multinomial, L, p, cluster_plate=-2)

# Generate random data from three clusters
# NOTE: This creates the data matrix as dense array first, so don't try to create huge datasets with this.
y = sparse.csr_matrix(
    bp.nodes.Multinomial(
        L,
        bp.nodes.Dirichlet(3e-2*np.ones(D), plates=(K_true,)).random()[np.random.randint(0, K_true, N)]
    )
    .random()
)

Y.observe(y)

# For comparison, you can try with dense data array
#Y.observe(y.toarray())

# Initialize soft cluster assignments randomly to break the symmetry in the model
Z.initialize_from_random()

# Run inference
Q = bp.inference.VB(Y, p, alpha, Z)
Q.update(p, Z, alpha, repeat=100)
```

I don't have very good visualization tools available. Below is some very simple ones. Please ask if you have problems examining the results as you'd like.


```python
%matplotlib notebook

# Some visualizations. These aren't very efficient, might not work (or might take a huge amount of time)
# for large variables

bp.plot.plt.figure()
bp.plot.hinton(Z, square=False)

bp.plot.plt.figure()
bp.plot.hinton(p[:,0], square=False) # need to get rid of the singleton plate axis because of bad support in Hinton
```

If you get an error that says "Requested sparse matrix sum-multiply not implemented", just give the error message details and I can probably easily add support, no problem.

I haven't thoroughly tested this feature, but at least it seemed to work and I got some speed improvements for large datasets. Please report if you notice anything weird or something doesn't work. It might be that, for instance, I've missed some spot where intermediate large dense arrays are created.

**Comment by dar326** [comment]
Awesome! Thanks. I will give this a try.

**Comment by dar326** [comment]
Sorry for the delay on providing feedback on this. It seems that with the changes to allow support for sparse matrices, I can no longer provide multinomial observations with a non-uniform number of trials. When specifying the mixture node representing my observation node, I provide the number of trials as an array when constructing the node. In the case that I use a dense array to run inference, this setup works. However, when I use a sparse matrix for my observations, I get the following error:

```
self = <bayespy.inference.vmp.nodes.multinomial.MultinomialDistribution object at 0x7f3e5a770c88>
x = <92370x256 sparse matrix of type '<class 'numpy.uint32'>'
        with 7127088 stored elements in Compressed Sparse Row format>
mask = True

    def compute_fixed_moments_and_f(self, x, mask=True):
        r"""
            Compute the moments and :math:`f(x)` for a fixed value.
            """
    
        # NOTE: In order to support sparse data, we must use some thin wrappers
        # over some of the NumPy functions
    
        # Check that counts are valid
        x = misc.asarray(x)
        if not misc.isinteger(x):
            raise ValueError("Counts must be integers")
        if misc.any(x < 0):
            raise ValueError("Counts must be non-negative")
        if np.any(np.sum(x, axis=-1) != self.N):
>           raise ValueError("Counts must sum to the number of trials")
E           ValueError: Counts must sum to the number of trials
```

Is there a way to set a non-uniform number of trials for the observation node or is that a casualty of providing this feature?

**Comment by jluttine** [comment]
Ok, good catch, thanks! I'm just going for a short holiday so I won't have time to look at this until next week some day. Sorry for the inconvenience.

**Comment by dar326** [comment]
No problem. Enjoy your holiday!

**Comment by jluttine** [comment]
Should be fixed now. In the same branch as before.

**Comment by dar326** [comment]
Thanks! Just letting you know that, using the updated code, the problem is no longer present.
